From 6b36f72f774df67bd1454be8b2aec628ded45ef0 Mon Sep 17 00:00:00 2001
From: Christian Schuster <christian.schuster@systec-electronic.com>
Date: Wed, 18 Dec 2024 16:25:49 +0100
Subject: [PATCH] drivers/thermal/k3_j72xx_bandgab.c: add sysfs support for
 thermal zones

Signed-off-by: Christian Schuster <christian.schuster@systec-electronic.com>
---
 drivers/thermal/k3_j72xx_bandgap.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/thermal/k3_j72xx_bandgap.c b/drivers/thermal/k3_j72xx_bandgap.c
index 0acafdfdd6f2..63a799ed7286 100644
--- a/drivers/thermal/k3_j72xx_bandgap.c
+++ b/drivers/thermal/k3_j72xx_bandgap.c
@@ -20,6 +20,8 @@
 #include <linux/delay.h>
 #include <linux/slab.h>
 
+#include "thermal_hwmon.h"
+
 #define K3_VTM_DEVINFO_PWR0_OFFSET		0x4
 #define K3_VTM_DEVINFO_PWR0_TEMPSENS_CT_MASK	0xf0
 #define K3_VTM_TMPSENS0_CTRL_OFFSET		0x300
@@ -515,6 +517,8 @@ static int k3_j72xx_bandgap_probe(struct platform_device *pdev)
 		}
 	}
 
+	devm_thermal_add_hwmon_sysfs(bgp->dev, ti_thermal);
+
 	platform_set_drvdata(pdev, bgp);
 
 	print_look_up_table(dev, ref_table);
-- 
2.47.0

