From 943b062cdac5ea5f090a716bf80aadd9b1556fcf Mon Sep 17 00:00:00 2001
From: Andreas Dinter <andreas.dinter@systec-electronic.com>
Date: Fri, 2 Aug 2024 15:34:25 +0200
Subject: [PATCH] board/ti/am62x/sysworxx.c: Remove obsolete board detection

Signed-off-by: Andreas Dinter <andreas.dinter@systec-electronic.com>
---
 board/ti/am62x/sysworxx.c | 94 ---------------------------------------
 1 file changed, 94 deletions(-)

diff --git a/board/ti/am62x/sysworxx.c b/board/ti/am62x/sysworxx.c
index bd916c1f392..a9214697b29 100644
--- a/board/ti/am62x/sysworxx.c
+++ b/board/ti/am62x/sysworxx.c
@@ -22,18 +22,9 @@
 #include <asm/gpio.h>
 #include <linux/delay.h>
 
-#include "../common/board_detect.h"
 #include "../common/fdt_ops.h"
-
 #include "../common/k3-ddr-init.h"
 
-#define board_is_am62x_skevm()  (board_ti_k3_is("AM62-SKEVM") || \
-				 board_ti_k3_is("AM62B-SKEVM"))
-#define board_is_am62b_p1_skevm() board_ti_k3_is("AM62B-SKEVM-P1")
-#define board_is_am62x_lp_skevm()  board_ti_k3_is("AM62-LP-SKEVM")
-#define board_is_am62x_sip_skevm()  board_ti_k3_is("AM62SIP-SKEVM")
-#define board_is_am62x_play()	board_ti_k3_is("BEAGLEPLAY-A0-")
-
 DECLARE_GLOBAL_DATA_PTR;
 
 #if CONFIG_IS_ENABLED(SPLASH_SCREEN)
@@ -64,8 +55,6 @@ int board_init(void)
 	return 0;
 }
 
-#if CONFIG_IS_ENABLED(TI_I2C_BOARD_DETECT)
-
 void reset_eth_phy(const char* phy_rst_name) {
 	int ret;
 	struct gpio_desc phy_rst_desc;
@@ -99,92 +88,9 @@ void reset_eth_phys(void) {
 	reset_eth_phy("gpio@601000_29");
 }
 
-int do_board_detect(void)
-{
-	int ret;
-
-	ret = ti_i2c_eeprom_am6_get_base(CONFIG_EEPROM_BUS_ADDRESS,
-					 CONFIG_EEPROM_CHIP_ADDRESS);
-	if (ret) {
-		printf("EEPROM not available at 0x%02x, trying to read at 0x%02x\n",
-		       CONFIG_EEPROM_CHIP_ADDRESS, CONFIG_EEPROM_CHIP_ADDRESS + 1);
-		ret = ti_i2c_eeprom_am6_get_base(CONFIG_EEPROM_BUS_ADDRESS,
-						 CONFIG_EEPROM_CHIP_ADDRESS + 1);
-		if (ret)
-			pr_err("Reading on-board EEPROM at 0x%02x failed %d\n",
-			       CONFIG_EEPROM_CHIP_ADDRESS + 1, ret);
-	}
-
-	return ret;
-}
-
-int checkboard(void)
-{
-	struct ti_am6_eeprom *ep = TI_AM6_EEPROM_DATA;
-
-	if (!do_board_detect())
-		printf("Board: %s rev %s\n", ep->name, ep->version);
-
-	return 0;
-}
-
-#if CONFIG_IS_ENABLED(BOARD_LATE_INIT)
-static void setup_board_eeprom_env(void)
-{
-	char *name = "am62x_skevm";
-
-	if (do_board_detect())
-		goto invalid_eeprom;
-
-	if (board_is_am62x_skevm())
-		name = "am62x_skevm";
-	else if (board_is_am62b_p1_skevm())
-		name = "am62b_p1_skevm";
-	else if (board_is_am62x_lp_skevm())
-		name = "am62x_lp_skevm";
-	else if (board_is_am62x_sip_skevm())
-		name = "am62x_sip_skevm";
-	else if (board_is_am62x_play())
-		name = "am62x_beagleplay";
-	else
-		printf("Unidentified board claims %s in eeprom header\n",
-		       board_ti_get_name());
-
-invalid_eeprom:
-	set_board_info_env_am6(name);
-}
-
-static void setup_serial(void)
-{
-	struct ti_am6_eeprom *ep = TI_AM6_EEPROM_DATA;
-	unsigned long board_serial;
-	char *endp;
-	char serial_string[17] = { 0 };
-
-	if (env_get("serial#"))
-		return;
-
-	board_serial = simple_strtoul(ep->serial, &endp, 16);
-	if (*endp != '\0') {
-		pr_err("Error: Can't set serial# to %s\n", ep->serial);
-		return;
-	}
-
-	snprintf(serial_string, sizeof(serial_string), "%016lx", board_serial);
-	env_set("serial#", serial_string);
-}
-#endif
-#endif
-
 #if CONFIG_IS_ENABLED(BOARD_LATE_INIT)
 int board_late_init(void)
 {
-	if (IS_ENABLED(CONFIG_TI_I2C_BOARD_DETECT)) {
-		setup_board_eeprom_env();
-		setup_serial();
-	}
-
-	ti_set_fdt_env(NULL, NULL);
 	reset_eth_phys();
 
 	return 0;
-- 
2.49.0

