From 0c15edf52339bcb55fd1e6090f2c5cb9303a3cba Mon Sep 17 00:00:00 2001
From: Andreas Dinter <andreas.dinter@systec-electronic.com>
Date: Fri, 2 Aug 2024 15:34:25 +0200
Subject: [PATCH 8/9] board/ti/am62x/sysworxx.c: Remove obsolete board
 detection

Signed-off-by: Andreas Dinter <andreas.dinter@systec-electronic.com>
---
 board/ti/am62x/sysworxx.c | 101 --------------------------------------
 1 file changed, 101 deletions(-)

diff --git a/board/ti/am62x/sysworxx.c b/board/ti/am62x/sysworxx.c
index e638beb45ad..2ce54b38db2 100644
--- a/board/ti/am62x/sysworxx.c
+++ b/board/ti/am62x/sysworxx.c
@@ -26,21 +26,12 @@
 
 #include <linux/sizes.h>
 
-#include "../common/board_detect.h"
 #include "../common/rtc.h"
 
 #include "../common/k3-ddr-init.h"
 
 DECLARE_GLOBAL_DATA_PTR;
 
-#define board_is_am62x_skevm()  (board_ti_k3_is("AM62-SKEVM") || \
-				 board_ti_k3_is("AM62B-SKEVM"))
-#define board_is_am62b_p1_skevm() board_ti_k3_is("AM62B-SKEVM-P1")
-#define board_is_am62x_lp_skevm()  board_ti_k3_is("AM62-LP-SKEVM")
-#define board_is_am62x_sip_skevm()  board_ti_k3_is("AM62SIP-SKEVM")
-#define board_is_am62x_play()	board_ti_k3_is("BEAGLEPLAY-A0-")
-#define board_is_am62x_sysworxx_ctr()	board_ti_k3_is("sysWORXX-CTR")
-
 #if CONFIG_IS_ENABLED(SPLASH_SCREEN)
 static struct splash_location default_splash_locations[] = {
 	{
@@ -171,101 +162,9 @@ void spl_perform_fixups(struct spl_image_info *spl_image)
 }
 #endif
 
-#ifdef CONFIG_TI_I2C_BOARD_DETECT
-int do_board_detect(void)
-{
-	int ret;
-
-	ret = ti_i2c_eeprom_am6_get_base(CONFIG_EEPROM_BUS_ADDRESS,
-					 CONFIG_EEPROM_CHIP_ADDRESS);
-
-	if (ret) {
-		printf("EEPROM not available at %d, trying to read at %d\n",
-			CONFIG_EEPROM_CHIP_ADDRESS, CONFIG_EEPROM_CHIP_ADDRESS + 1);
-		ret = ti_i2c_eeprom_am6_get_base(CONFIG_EEPROM_BUS_ADDRESS,
-						 CONFIG_EEPROM_CHIP_ADDRESS + 1);
-		if (ret)
-			pr_err("Reading on-board EEPROM at 0x%02x failed %d\n",
-				CONFIG_EEPROM_CHIP_ADDRESS + 1, ret);
-	}
-	return ret;
-}
-
-int checkboard(void)
-{
-	struct ti_am6_eeprom *ep = TI_AM6_EEPROM_DATA;
-
-	if (!do_board_detect())
-		printf("Board: %s rev %s\n", ep->name, ep->version);
-
-	return 0;
-}
-#endif
-
 #ifdef CONFIG_BOARD_LATE_INIT
-static void setup_board_eeprom_env(void)
-{
-	char *name = "am62x_skevm";
-
-	if (do_board_detect())
-		goto invalid_eeprom;
-
-	if (board_is_am62x_skevm())
-		name = "am62x_skevm";
-	else if (board_is_am62b_p1_skevm())
-		name = "am62b_p1_skevm";
-	else if (board_is_am62x_lp_skevm())
-		name = "am62x_lp_skevm";
-	else if (board_is_am62x_sip_skevm())
-		name = "am62x_sip_skevm";
-	else if (board_is_am62x_play())
-		name = "am62x_beagleplay";
-	else if (board_is_am62x_sysworxx_ctr())
-		name = "am62x_sysworxx_ctr";
-	else
-		printf("Unidentified board claims %s in eeprom header\n",
-		       board_ti_get_name());
-
-invalid_eeprom:
-	set_board_info_env_am6(name);
-}
-
-static void setup_serial(void)
-{
-	struct ti_am6_eeprom *ep = TI_AM6_EEPROM_DATA;
-	unsigned long board_serial;
-	char *endp;
-	char serial_string[17] = { 0 };
-
-	if (env_get("serial#"))
-		return;
-
-	board_serial = simple_strtoul(ep->serial, &endp, 16);
-	if (*endp != '\0') {
-		pr_err("Error: Can't set serial# to %s\n", ep->serial);
-		return;
-	}
-
-	snprintf(serial_string, sizeof(serial_string), "%016lx", board_serial);
-	env_set("serial#", serial_string);
-}
-
 int board_late_init(void)
 {
-	if (IS_ENABLED(CONFIG_TI_I2C_BOARD_DETECT)) {
-		struct ti_am6_eeprom *ep = TI_AM6_EEPROM_DATA;
-
-		setup_board_eeprom_env();
-		setup_serial();
-		/*
-		 * The first MAC address for ethernet a.k.a. ethernet0 comes from
-		 * efuse populated via the am654 gigabit eth switch subsystem driver.
-		 * All the other ones are populated via EEPROM, hence continue with
-		 * an index of 1.
-		 */
-		board_ti_am6_set_ethaddr(1, ep->mac_addr_cnt);
-	}
-
 	reset_eth_phys();
 
 	return 0;
-- 
2.46.0

